### PR TITLE
Bugfix: Filesystem object not passed to pyarrow reader

### PR DESCRIPTION
**Pull request by fjetter**
This fixes a bug where the FileSystem object is not properly passed to the `read_parquet` method.

I had to add additional methods to the `LocalFileSystem` class since `pyarrow` expects a broader interface. These adaptations should eventually also be included in s3fs to support `pyarrow` reads from s3.

### CODE REVIEW

**Comment by wesm** [comment]
Thanks @fjetter -- I started a bit of refactoring recently to enable the Dask filesystems (s3fs, gcsfs, etc.) to be used in `ParquetDataset` (see ARROW-1213). 

This patch helps triage use in Dask right now, but the main thing that needs to change is to use the `walk` API on the filesystems rather than a combination of ls, isdir, and isfile. Some filesystems like S3 make it a little bit awkward to determine `isfile` or `isdir` and it's simpler and faster to emulate `walk`.

File: dask/bytes/local.py [hunk 1]
```diff
@@ -26,6 +26,10 @@ def __init__(self, **storage_options):
         """
         self.cwd = os.getcwd()
 
+    @property
+    def pathsep(self):
+        return '/'
```
Comment (mrocklin) [inline dask/bytes/local.py]: Perhaps `os.path.sep`?

Comment (martindurant) [inline dask/bytes/local.py]: `LocalFileSystem.sep` is exactly that.

**Comment by martindurant** [comment]
In favour of using walk (i.e., produce a set of files within a location - not the python iterator) as, indeed, "isdir" is ill-defined for s3 and gcs (which both happen to have `walk` defined already).

**Comment by fjetter** [comment]
I agree that `isdir`, `isfile`, etc. is rather weird for s3 and I was already wondering how one would do that. 
How do you propose to continue? Include `walk` to the local filesystem and wait for https://issues.apache.org/jira/browse/ARROW-1213 to enable `pyarrow`?

**Comment by wesm** [comment]
If adding isdir/isfile to Dask's local FS enables directory reads in Dask to work, then that is helpful for testing. I will see if I can get a patch up for ARROW-1213 today or tomorrow -- it will miss the 0.5.0 release but I suspect we have some more testing and hardening to do to make sure that we have all the edge cases for S3 / non-local filesystems sorted out. If you could help with "kicking the tires" that would be much appreciated

We're looking to do 0.6.0 within any luck within a few weeks

cc @yackoa who was also asking me about this

**Comment by fjetter** [comment]
Sure, once you've got a working prototype I should be able to try it out, at least for S3.

**Comment by wesm** [comment]
@fjetter this should be working in Arrow 0.6.0 (release candidate for this should be out next week) -- a version of this patch is necessary to pass the filesystem to `ParquetDataset`. Can you take a look at testing this and then we can merge a patch to Dask after 0.6.0 is out?

**Comment by fjetter** [comment]
`pyarrow` still rejects the dask `LocalFilesystem` since [this](https://github.com/apache/arrow/blob/0b91cadfe1cdf5964f275c7d8324332e9e49b0d9/python/pyarrow/parquet.py#L646) check is too restrictive which is why the tests are still failing. Although I didn't check yet, I assume the `DaskS3Filesystem` should be rejected as well. I opened https://issues.apache.org/jira/browse/ARROW-1417 to track this issue

**Comment by wesm** [comment]
That's a bummer. I never got any feedback from my comment a few weeks ago so would have been nice to fix this before 0.6.0 went out. I need some help with integration testing with Dask (a Dockerfile in the Arrow codebase would be sufficient) so that we can keep this integration stable and working. 

0.7.0 is likely ~2 weeks away so let's resolve this before then.

**Comment by wesm** [comment]
cc @jreback @cpcloud in case you have any spare cycles to help

**Comment by fjetter** [comment]
Tests should become green once https://github.com/apache/arrow/pull/1032 is merged and released. 

I encountered two issues, though. 
1. I was running the tests on python 2.7 and encountered [this](https://github.com/dask/fastparquet/issues/145) issue and could only fix it by using unicode literals. This might be an actual `fastparquet` bug after all
2. Setting the index of the pandas data frame explicitly doesn't feel natural to me since I'd expect arrow to perform this action for me. I have the impression that I stumbled either upon an arrow bug or am misusing it somehow. Does anybody know which one it is?

Sorry for taking so much time to fix this!

**Comment by wesm** [comment]
If the schema metadata is being written and read properly, you should not have to set the pandas index explicitly. So we should track down the origin of the bug. I think @jreback has also done some integration testing of this in pandas

**Comment by fjetter** [comment]
After https://github.com/dask/dask/pull/2822 was merged this reduces to a minimal fix. The tests seem to break for the same reason master is red

**Comment by fjetter** [comment]
cc @mrocklin anything missing?

**Comment by TomAugspurger** [comment]
@martindurant, any thoughts?

I'll merge this later today otherwise.

**Comment by martindurant** [comment]
OK with me

File: dask/bytes/tests/test_s3.py [hunk 2]
```diff
@@ -283,9 +283,17 @@ def test_read_text_passes_through_options():
         assert df.count().compute(get=get) == 3
 
 
-def test_parquet(s3):
+@pytest.mark.parametrize("engine", [
+    'auto',
+    'pyarrow',
```
Comment (jcrist) [inline dask/bytes/tests/test_s3.py]: In the current version the dask engine keyword is `'arrow'`, not `'pyarrow'`. This is fixed in #2868, but for now you'll need to change the engine name.

**Comment by jcrist** [comment]
> Moving boto3 and moto to install from conda-forge should fix this.

PR: https://github.com/dask/dask/pull/2875

**Comment by jcrist** [comment]
I pushed a fix for the test here. Once tests pass will merge. Thanks @fjetter.
